sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -702,7 +705,7 @@ class Dataset[T] private[sql](
    * @since 2.0.0
    */
   def join(right: Dataset[_]): DataFrame = withPlan {
-    Join(logicalPlan, right.logicalPlan, joinType = Inner, None)
+    Join(AnalysisBarrier(logicalPlan), right.logicalPlan, joinType = Inner, None)
```

Comment:
I am wondering if we should check there's duplication between right and left sides and decide using barrier or not for right side.